esp-hal/src/spi/master.rs

```diff
@@ -362,6 +361,155 @@ impl Address {
     }
 }
 
+pub struct WholeFifo;
+pub struct HalfFifo {
+    is_high_part: bool,
+}
+
+pub trait FifoFraction: crate::private::Sealed {
+    const SIZE: usize;
+
+    fn range(&self) -> Range<usize>;
+}
+
+impl crate::private::Sealed for WholeFifo {}
+
+impl FifoFraction for WholeFifo {
+    #[cfg(not(esp32s2))]
+    const SIZE: usize = 64;
+    #[cfg(esp32s2)]
+    const SIZE: usize = 72;
+
+    fn range(&self) -> Range<usize> {
+        0..Self::SIZE
+    }
+}
+
+impl crate::private::Sealed for HalfFifo {}
+
+impl FifoFraction for HalfFifo {
+    const SIZE: usize = WholeFifo::SIZE / 2;
+
+    fn range(&self) -> Range<usize> {
+        if self.is_high_part {
+            Self::SIZE..WholeFifo::SIZE
+        } else {
+            0..Self::SIZE
+        }
+    }
+}
+
+pub struct Fifo<'d, T, S> {
```

Comment:
I don't know if `Fifo` is the most appropriate name for this. The TRM really refers to this as the "CPU-Controlled Buffer", and the "FIFO" really refers to the internal buffer in the peripheral that is inaccessible to the CPU except via reset bits.
I'll leave this bike-shedding for later.

Reply:
Maybe just `SpiBuf` or `SpiSwap`?

Reply:
I like `SpiBuf`

Reply:
Updated